File: TopCoder06/Ruby/Main.rb
class FriendScore
  def highestScore(friends)
    ans = 0
    n = friends[0].length
    for i in 0...n
      cnt = 0
      for j in 0...n
        next if i == j
        if friends[i][j] == 'Y'
          cnt+=1
        else
          for k in 0...n
            if friends[j][k] == 'Y' && friends[k][i] == 'Y'
              cnt+=1
              break
            end
          end
        end
      end
      ans = [ans,cnt].max
    end
    return ans
  end
end

friends = ["NYNNN","YNYNN","NYNYN","NNYNY","NNNYN"]
puts FriendScore.new.highestScore(friends)

File: TopCoder06/PHP/Main.php
<?php

class FriendsScore {
	public function highestScore($friends) {
		$ans = 0;
		$n = strlen($friends[0]);
		for ($i = 0; $i < $n; $i++) {
			$cnt = 0;
			for ($j = 0; $j < $n; $j++) {

				if ($i == $j) {
					continue;
				}
				if ($friends[$i]{$j} == 'Y') {
					$cnt++;
				} else {
					for ($k = 0; $k < $n; $k++) {
						if ($friends[$j]{$k} == 'Y' && $friends[$k]{$i} == 'Y') {
							$cnt++;
							break;
						}
					}
				}
			}
			$ans = max($ans, $cnt);
		}
		return $ans;
	}

}

$friends = array("NYNNN", "YNYNN", "NYNYN", "NNYNY", "NNNYN");
$fs = new FriendsScore;
echo $fs -> highestScore($friends);
?>